### PR TITLE
character preview performance pass

### PR DESCRIPTION
**Pull request by BenLubar**
## About the PR

Currently looks like this:

image

Takes about an extra frame for it to update position when scrolling or resizing the window.

## Why's this needed?

Performance

### CODE REVIEW

**Comment by BenLubar** [comment]
The character setup preview works great, but I tried doing a similar thing to the clothing booth and it didn't work at all. Not sure what the difference is.

**Comment by BenLubar** [comment]
Could probably benefit from a background texture or color on the preview map.

File: code/datums/character_preview.dm [hunk 1]
```diff
@@ -0,0 +1,119 @@
+datum/character_preview
```
Comment (ZeWaka) [inline code/datums/character_preview.dm:1]: some dmdocumentation would be great on this

File: code/datums/character_preview.dm [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+				"title" = "Character Preview",
+				"can-close" = FALSE,
+				"can-resize" = FALSE,
+			)))
+
+			. = ..(viewer, winid)
+
+		disposing()
+			. = ..()
+			SPAWN_DBG(0)
```
Comment (ZeWaka) [inline code/datums/character_preview.dm:26]: needed?

File: code/datums/character_preview.dm [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+				viewer.screen -= src.preview_mob
+
+		proc/RemoveAllClients()
+			for (var/client/viewer in src.viewers)
+				if (viewer)
+					viewer.screen -= src.handler
+					viewer.screen -= src.preview_mob
+			src.viewers.len = 0
+
+	New(client/viewer, window_id, control_id = null)
```
Comment (ZeWaka) [inline code/datums/character_preview.dm:65]: this is really gross-ly relatively pathed, with this being below all of the children and their News

File: code/datums/character_preview.dm [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+		var/mob/living/carbon/human/H = new()
+		src.preview_mob = H
+		H.screen_loc = "[src.preview_id];1,1"
+		src.handler.vis_contents += H
+		src.viewer?.screen += H
+
+	disposing()
+		STOP_TRACKING
+		SPAWN_DBG(0)
```
Comment (ZeWaka) [inline code/datums/character_preview.dm:97]: needed?

File: code/datums/character_preview.dm [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	proc/update_appearance(datum/appearanceHolder/AH, datum/mutantrace/MR = null, direction = SOUTH)
+		src.preview_mob.dir = direction
+		src.preview_mob.set_mutantrace(null)
+		src.preview_mob.bioHolder.mobAppearance.CopyOther(AH)
+		src.preview_mob.set_mutantrace(MR)
+		src.preview_mob.organHolder.head.donor = src.preview_mob
+		src.preview_mob.organHolder.head.donor_appearance.CopyOther(src.preview_mob.bioHolder.mobAppearance)
+		src.preview_mob.update_colorful_parts()
+		src.preview_mob.set_body_icon_dirty()
+		src.preview_mob.set_face_icon_dirty()
```
Comment (ZeWaka) [inline code/datums/character_preview.dm:119]: there's no wrapper for all this shit? @Superlagg?

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/datums/character_preview.dm (hunk 2 above)
Comment (BenLubar) [reply code/datums/character_preview.dm:26]: It was in the viewport code, and I copied it from there.

File: code/datums/character_preview.dm (hunk 4 above)
Comment (BenLubar) [reply code/datums/character_preview.dm:97]: It was in the viewport code, and I copied it from there.

File: code/datums/character_preview.dm (hunk 5 above)
Comment (Superlagg) [reply code/datums/character_preview.dm:63]: Nope! Don't know what that is.

File: code/datums/character_preview.dm (hunk 1 above)
Comment (ZeWaka) [reply code/datums/character_preview.dm:1]: see above

**Review by ZeWaka — APPROVED** [review]

**Pull request merged**